File: car_iot_android/app/src/main/kotlin/info/karuru/cariot/meter/MeterSlot.kt
package info.karuru.cariot.meter

import info.karuru.cariot.obd.GaugeStyle
import info.karuru.cariot.obd.ObdMetric
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

// メーター画面の1タイル（表示項目＋ゲージ種別）。mobile/lib/models/meter_slot.dart相当。
@Serializable
data class MeterSlot(val metric: ObdMetric, val style: GaugeStyle)

// MeterConfigStore(永続化)・インポート/エクスポートの両方から使うJSON変換ロジック。
// 永続化層から切り離すことでテスト容易にする（upload/ObdUploader.ktのbuildRequestBodyと
// 同じ設計判断、Phase8後半）。
object MeterSlotJson {
  private val json = Json { ignoreUnknownKeys = true }

  fun encode(slots: List<MeterSlot>): String = json.encodeToString(slots)

  // 壊れたJSONはnullを返す（呼び出し側でデフォルト値へのフォールバックを判断する）。
  fun decode(raw: String): List<MeterSlot>? {
    return try {
      json.decodeFromString(raw)
    } catch (e: Exception) {
      null
    }
  }
}

File: mobile/lib/models/meter_slot.dart
import 'obd_metric.dart';

// メーター画面の1タイル分の設定（どの項目をどの形式で表示するか）。
// 同一metricを複数タイルで選ぶことも許可する（禁止ロジックは持たせない）。
class MeterSlot {
  final ObdMetric metric;
  final GaugeStyle style;

  const MeterSlot({required this.metric, required this.style});

  Map<String, String> toJson() => {'metric': metric.name, 'style': style.name};

  // 対応するenum値が見つからない場合（将来のリネーム等）はnullを返す。
  // 呼び出し側で読み捨てられるようにし、アプリをクラッシュさせない。
  static MeterSlot? fromJson(Map<String, dynamic> json) {
    ObdMetric? metric;
    for (final m in ObdMetric.values) {
      if (m.name == json['metric']) {
        metric = m;
        break;
      }
    }
    GaugeStyle? style;
    for (final s in GaugeStyle.values) {
      if (s.name == json['style']) {
        style = s;
        break;
      }
    }
    if (metric == null || style == null) return null;
    return MeterSlot(metric: metric, style: style);
  }
}
